### ISSUE TITLE
[RFC] [Doc] Aurelia 2 routing compared to v1 (and other frameworks)

### ISSUE THREAD

**Issue by jwx** [aurelia/aurelia#582]
# 💬 RFC

This document describes how the routing in Aurelia v2 differs from the routing in v1 (and other frameworks which uses a url-based routing). If you're new to client side routing,  [Routing with Aurelia](https://github.com/aurelia/aurelia/issues/369) might be a better starting point.

## Two routers in one

In addition to having a new way of dealing with routing, the v2 router also supports the more traditional, url-based routing of v1. This document will however focus solely on the new component-based routing.

# New mental model

The main difference between v1 and v2 routing is that in v1 you _compose a url string_ to tell the router, via an alias (a configured route), which components you want to load. In v2 you tell the router directly _which components to load_ and the router will take care of the rest, including creating the url string.

The url-based routing of v1 requires some kind of configuration to connect url strings to sets of components in viewports. The component-based routing of v2 doesn't require any "connect configuration" since we're working directly with the end result: which components to load.

<details><summary>View code example</summary>

### v1
`app.ts` - configure the route/url string `users` to load `Users` component
```typescript
public configureRouter(config: RouterConfiguration, router: Router) {
  config.map([
    { route: 'users', name: 'users', moduleId: './components/users', },
  ]);
}
```
`app.html` - link with route/url string above
```html
<a href="#/users">Show users</a>
<router-view name="main"></router-view>
```

### v2
`app.ts` - nothing is needed in `app.ts`

`app.html` - link directly with the name of the component to load
```html
<a href="users">Show users</a>
<au-viewport name="main"></au-viewport>
```
<details><summary><b>How does the router know where the component is?</b><br><br>
</summary>

In Aurelia 2, all component dependencies must be registered either globally or on the component that has the dependency. This is a general principle and not just something for routing. The router uses the dependency registration to match component names to components. The actual location, the file, of the component is therefore never needed. You can read more about registering dependencies <a href="#no-link-yet">here</a>.
</details>
</details>

## Different names

While we've tried to avoid breaking changes, the differences in behavior have merited some name changes.

<details><summary>View list of commonly used names that have changed<br><br>
</summary>

In html
```
router-view  ->  au-viewport
```

In javascript/typescript
```
canActivate    ->  canEnter
activate       ->  enter
canDeactivate  ->  canLeave
deactivate     ->  leave
navigate       ->  goto
navigateTo     ->  goto
```
</details>

**Comment by davismj** [comment]
> `href         ->  au-href`

Might be confusing. It really hasn't changed. And route-href isn't au-href either. So maybe leave this part off for now.

**Comment by jwx** [comment]
@davismj You're right; removed it. Thanks!

**Comment by arnederuwe** [comment]
While I really like the new v2 mental model for it's simplicity, We used the old configuration model for more than just pointing route X to component X, 3 things from the top of my head:
- It's also the place where we would define stuff like authorization. Most of the time we would do something like this:
```
{
	route: ["task-update"],
	moduleId: "app/tasks/task-update",
	title: "Edit task",
	name: "task-update",
	settings: {
		roles: [Roles.Register, Roles.Planner, Roles.Administrator]
	},
},
```
With the new model, where would you put stuff like this? One solution would be to decorate your components with this security metadata, which seems fine. But you would still lose the flexibility of defining 2 routes pointing to the same component, where the authorization of route 1 differs from route 2 (for example, a create route and an update route, pointing to the same component)
- You could tell the framework to execute a different activation strategy
```
{
	route: ["task-update"],
	moduleId: "app/tasks/task-update",
	title: "Aanpassen",
	name: "task-update",
	activationStrategy: activationStrategy.replace,
},
```
The same argument as above, you can also configure this on your component but you would be out of luck when you want to do one or the other depending on a specific route
- It was easy to build a navbar from routeconfig, you could just iterate the config and build a navbar, throw in some custom properties to show/hide or build some hiërarchy. Will this be possible in this model? Perhaps by doing some AOT magic, building metadata about each route pointer in the app? Doesn't seem easy to pull off (to me :) )

I would like to repeat I really like the new model, but we need at least 2 of the things I mentioned on each project we do, so I'm interested in hearing your opinions on how you would tackle these scenario's in the new model.

**Comment by fkleuver** [comment]
The old model will also still be available, and to a certain extent the two will be usable in conjunction. This might be one of those cases where that would be the way to go.

**Comment by arnederuwe** [comment]
> The old model will also still be available, and to a certain extent the two will be usable in conjunction. This might be one of those cases where that would be the way to go.

I am fully aware of that, but my point implies we can never use the new model if stuff like this isn't possible, and I'm quite sure many others have the same requirements.

Also, I am looking into making the [aurelia openid connect plugin](https://github.com/aurelia-contrib/aurelia-open-id-connect) compatible with V2, and right now I don't see a way to incorporate the new router into it. It would be a shame to require users to always use the configuration model if they want to use the OIDC plugin IMO

**Comment by fkleuver** [comment]
What I meant by

> and to a certain extent the two will be usable in conjunction

Is that you could do stuff like this:

```ts
@route({ settings: { roles: [Roles.Register, Roles.Planner, Roles.Administrator] } })
export class TaskUpdate {

}
```
```ts
@route({ activationStrategy: activationStrategy.replace })
export class TaskUpdate {

}
```

To associate additional settings with a particular component via configuration, while relying on the new model to figure out the rest (path, name, etc) by convention.

The general idea behind the new model is that most configurable properties can be specified in the path / instruction without having to configure them. For some configurable properties this is either tricky to do, or for various reasons just not desirable (e.g. you might want to keep the url clean / looking more like traditional urls).

To address this point:

> But you would still lose the flexibility of defining 2 routes pointing to the same component, where the authorization of route 1 differs from route 2 (for example, a create route and an update route, pointing to the same component)

This could be addressed in several ways. The first one, being to define this on the component itself with two separate route configs:

```ts
@route({ path: ':id', settings: { roles: [Roles.Editor] } })
@route({ path: 'new', settings: { roles: [Roles.Anonymous] } })
export class EditThing {

}
```

This is an example of the old and new model working in conjunction. Only the `path` is specified along with custom settings to be able to differentiate the two, but this is on the component itself. There is still no configuration on the "parent" side.

That said, there could be configuration on the "parent" side regardless, even combined with some config on the component itself:
```ts
{
  path: ':id',
  settings: { roles: [Roles.Editor] },
  component: EditThing,
},
{
  path: 'new',
  settings: { roles: [Roles.Anonymous] }
  component: EditThing,
},

// ...

// Not differentiated by path, so applicable to all paths
@route({ activationStrategy: activationStrategy.replace })
export class EditThing {

}
```

**Comment by jwx** [comment]
@arnederuwe The new router comes with _routing hooks_. There'll be several of them, covering different parts of "centralized configuration". Currently, the api is
```typescript
router.addHook(callbackMethod, options);
```
where `callbackMethod` is a method you provide with the logic you want for the specified hook. Parameters into and return value from the callback method varies depending on which hook type it is. The options in the `options` parameter also varies depending on hook type, but one is consistent: `type`, that specifies which hook type it is.

So, for your first two use cases I'd say the `BeforeNavigation` hook could be useful. For example
```typescript
  this.router.addHook((instructions: NavigationInstruction[]) => 
  {
    for (const instruction of instructions) {
      if (!this.verifyAccess(instruction.componentName)) {
        return false; // Prevents the navigation, it's also possible to redirect
      }
    }
    return true; // Proceed
  }, 
  {
    type: HookTypes.BeforeNavigation, // Default, can be omitted in this case
    exclude: [ '', 'login' ], // Exclude public components from this hook. There's also an `include` option
  });

checkAccess(component) {
  const requirements = {
    'task-view': [Roles.View],
    'task-update': [Roles.Register, Roles.Planner, Roles.Administrator],
  }[component];
  for (const requirement of requirements) {
    if (user.roles.some(role => role === requirement)) {
      return true;
    }
  }
  return false;
}
```
Okay, pretty ugly code, but hopefully it illustrates the possibilites regarding centralized configuration.

As for the navigation, any list structure will suffice to iterate through. 
```html
<nav>
  <ul>
    <li repeat.for="link of [{ link: 'tasks', title: 'Tasks' }, { link: 'closed-tasks', title: 'Closed tasks' }]">
      <a goto="${link.link}">${link.title}</a>
    </li>
  </ul>
</nav>
```
There's also an `au-nav` custom element that lets you do this in code
```html
<au-nav name="tasks-menu"></au-nav>
```
```typescript
  router.setNav('tasks-menu', [{ component: 'tasks', title: 'Tasks' }, { component: 'closed-tasks', title: 'Closed tasks' }]);
```
The `au-nav` also supports child menues (`children`) and styling options. You might want to keep full control for complex menues and use cases, but for basic use cases it's a way to quickly and easily get a navbar (or tab list!) going.

Bottom line is that the goal is that if you're able to configure it centrally in v1, you should be able to do it with the new routing approach as well (where it makes sense).

**Comment by arnederuwe** [comment]
Thank you @fkleuver and @jwx !
You provided me with everything I need in order to start putting something together, it seems I was missing some critical information about the new model, all cleared up now :)

**Comment by fkleuver** [comment]
To be clear, the APIs I mentioned don't exist yet. That's what I'll be working on towards the end of this month and beginning of January.

**Comment by arnederuwe** [comment]
As mentioned on Discord, That's fine :) 
Give me a ping when this is ready testdrive

**Comment by sspilleman** [comment]
Im also trying to put some routing example together, but it seems none of the is already available? Or am I looking in the wromg place?

**Comment by ShawnTalbert** [comment]
I've never liked how popular routers required me to manually specify url strings in config files and then remember and refer to them in HTML.  Those urls had to be maintained manually and generally had no support for intellisense or strong typing for correctness in IDEs.

I think the new approach is a good step forward (or a step towards what non-web development had in early 2000's  :smiley:  )

**Comment by jwx** [comment]
Some live coded routing can be found [here](https://www.youtube.com/playlist?list=PLWhGw6UrvSU5l3jY_89_ainf9Ofxy39FF).

**Comment by thomas-darling** [comment]
What is the recommended way to implement lazy-loading with this new router?

In v1, components are referenced as e.g. `PLATFORM.moduleName("./pages/foo", "foo")`, where the second parameter is the chunk name - and then that chunk would not be fetched from the server until the user actually navigates to the route.

In v2, components are referenced as e.g. `import("./pages/foo")` in the `@route` decorator, which means they are all fetched immediately when the app starts - which is a pretty big performance problem in a large app.

**Comment by jwx** [comment]
@thomas-darling In the direct router you can use `() => import("./pages/foo")` in the `@routes` decorator or in the `static routes` property for lazy loaded routes. Lazy loaded routes are on their way to the default router very soon as well.

**Comment by bigopon** [comment]
Lazy loaded route is in the default router. You can do any of the following
```ts
// 1
component: () => import('...')

// 2
component: import('...')

// 3
import { MyComponent } from '...'
component: MyComponent
```

**Comment by thomas-darling** [comment]
What is recommended way to dynamically configure routes and child routes at runtime?

In v1 we had the `configureRouter` method in each component, where we could manually configure the routes for the router associated with each component - and potentially do so based on conditions or data that is only available at runtime.

Some examples of scenarios where this may be required:
* You are building a webshop, and need to dynamically configure routes based on categories and products in your database.
* You are using a CMS, and need to dynamically configure routes based on the pages stored there.
* You need to dynamically configure routes e.g. depending on country, language, domain, tenant, active A/B tests, etc.

In v2 it seems that all route configuration must now be _statically defined at build time_, in decorators or static properties - with no way to dynamically configure the routes associated with a component. If this understanding is correct, it would  effectively make Aurelia unusable for a wide range of common scenarios - and make upgrading of some of our v1 apps pretty much impossible.

Hopefully I'm just missing something here, as everything else is looking pretty great so far 🙂 
Is there a plan to support runtime route configuration, equivalent to the `configureRouter` method in v1?

**Comment by jwx** [comment]
In the Aurelia Direct Router, dynamic/runtime routing is/will be possible primarily in two ways:
1) by using direct routing (since it's doesn't have configured routes but instead resolves the routes during runtime), and
2) by using the `router.configuration.addRoutes(configuredRoutes, contextViewModel)` method (which is currently unavailable in the latest release, but it will soon make a comeback).

(Note that actually making components available is outside of the router's responsibility.)

**Comment by thomas-darling** [comment]
Thanks, sounds good 🙂 

I did some more digging based on option 2 - would it be correct to assume, it will look something like this?

```ts
@inject
export class App {

  public constructor(@IRouter router: IRouter) {
      this.router = router;
  }

  private readonly router: Router;

  public binding() {
    this.router.addRoutes([
      {
        path: "authors/:id",
        instructions: [{ component: "author", viewport: "left" }],
        settings: { /* custom settings, such as a list of required permissions */ }
      },
      // etc.
    ]);
  }

  public onButtonClicked() {
    this.router.goto("authors/123"); // navigate to URL
  }
}
```

A couple of follow up questions:

- Components appear to be referenced by name, but how does that work, exactly?
  
  If I understand this correctly, the component name specified here, is the name of a custom element, which would need to be registered in the DI container. But that raises the question of when that registration needs to occur, and how this would work, if we want to lazy-load the component, when the user navigates to the route?

  I was expecting to see something like `component: () => import("./pages/author")` here - is that supported?
  
- In the example, I included a `settings` property in the route config, like we had in v1.
  
  Is it correct to assume this will be supported? - meaning that we could somehow intercept the navigation, inspect the settings associated with every route config involved in the navigation, and then based on those, determine e.g. whether the user is authorized to access the route?

- In v1 we had child routers, where routes configured in a child router would effectively be prefixed with the parent route.
  
  Is it correct to assume this behavior will remain the same? - meaning that if we also inject an `IRouter` into the `author` component referenced in the example, that component could itself add a route to a `book` component, with the path `books/:id`, such that the full URL for a book would be e.g. `authors/123/books/456`?

Finally, if you have any pointers to additional documentation about where this is going, it would be very much appreciated 🙂

**Comment by jwx** [comment]
Your assumptions about the code are correct except that `settings` is `data` and `goto` should be `load`.

- When a component is to be loaded/navigated to, the direct router will resolve the specified name, component, promise for a component (`import`) or function returning any of the previous. So

`component: 'author'` will try and find the named component in DI container 
`component: Author` will load the component
`component: import('./author')` will (if needed) wait for the promise to resolve and load the component
`component: () => import('./author')` will run the function and act according to above on the result
- The context (which can be either a view model or a DOM element) indicates where the routes should be added. It's more or less the equivalent of which component class the decorator/static property is added to. Your usage of `this` is probably going to be the most common one.
- The recommended property for your data will be `data`, but nothing is really preventing you from using `settings`.
- Yes, configured (and direct) routes are aware of their routing context and relative routes, as in not starting with `/`, will be added to their context (which is why it should be provided with `addRoutes`).

Sadly, I've got to say that I haven't got any good docs about this available yet. (In a way, the above pretty much covers most of it.) Please however feel free to ask questions until the docs are in place.

**Comment by thomas-darling** [comment]
Sounds great, thank you! 🙂 

Regarding docs, I agree, this pretty much covers everything - at least everything I can think of right now.
I mostly just asked, in case there were some caveats I wasn't aware of - but so far, this looks really good.

One thing though - in your previous comment you wrote:
> (Note that actually making components available is outside of the router's responsibility.)

At first, I thought that was probably a reference to the import/registration stuff, but if all that lazy loading goodness is already supported, I'm not quite sure what you meant by this. If you could just clarify that one, it would be awesome 🙂

**Comment by Sayan751** [comment]
@Vheissu @jwx Is this issue still relevant? Considering the router docs are already in place?

**Comment by Vheissu** [comment]
Correct, @Sayan751, this was during the preliminary router planning stages.
